libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotExternalEventsHandler.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.Bot.Builder.Adapters;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+using System;
+using System.IO;
+using System.Linq;
+using System.Security.Claims;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Integration.AspNet.Core.Handlers
+{
+    public class BotExternalEventsHandler : BotMessageHandlerBase
+    {
+        public BotExternalEventsHandler()
+        {
+        }
+
+        protected override async Task<InvokeResponse> ProcessMessageRequestAsync(HttpRequest request, BotFrameworkAdapter botFrameworkAdapter, Func<ITurnContext, Task> botCallbackHandler, CancellationToken cancellationToken)
+        {
+            ClaimsIdentity botIdentity = GetBotIdentityFromRequest();
+            Activity eventActivity = GetEventActivityFromRequest();
+
+            await botFrameworkAdapter.ProcessActivityAsync(botIdentity, eventActivity, botCallbackHandler, cancellationToken);
+
+            return null;
+
+            ClaimsIdentity GetBotIdentityFromRequest()
+            {
+                const string BotAppIdHttpHeaderName = "MS-BotFramework-BotAppId";
+                const string BotIdQueryStringParameterName = "BotAppId";
+
+                if (!request.Headers.TryGetValue(BotAppIdHttpHeaderName, out var botAppIdHeaders))
+                {
+                    if (!request.Query.TryGetValue(BotIdQueryStringParameterName, out botAppIdHeaders))
+                    {
+                        throw new InvalidOperationException($"Expected a Bot App ID in a header named \"{BotAppIdHttpHeaderName}\" or in a querystring parameter named \"{BotIdQueryStringParameterName}\".");
+                    }
+                }
+
+                var botAppId = botAppIdHeaders.First();
+
+                var result = new ClaimsIdentity("Bot-ExternalEvent-Auth");
+                result.AddClaim(new Claim(AuthenticationConstants.AudienceClaim, botAppId));
+                result.AddClaim(new Claim(AuthenticationConstants.AppIdClaim, botAppId));
+
+                return result;
+            }
```

Comment:
this feels like duplicate code from somewhere.  Is there a better way to refactor this?

Reply:
Yes, it is indeed similar in that it's building up a specifically "shaped" `ClaimsIdentity`, but it's totally different in terms of _where_ it gets the values for those claims from. The difference is between the bot identify coming in through in the JWT via the "regular" activity endpoint vs being specified in either a header or querystring in the case of this "external events" endpoint.